### PR TITLE
Feature/trained models inference

### PR DESCRIPTION
**Pull request by djaniak**

### CODE REVIEW

File: embeddings/model/lightning_module/huggingface_module.py [hunk 1]
```diff
@@ -24,11 +25,11 @@ def __init__(
         self.save_hyperparameters({"downstream_model_type": downstream_model_type.__name__})
         self.downstream_model_type = downstream_model_type
         self.config_kwargs = config_kwargs if config_kwargs else {}
+        self.configure_model()
```
Comment (ktagowski) [inline embeddings/model/lightning_module/huggingface_module.py:28]: Why is it moved to `__init__`?

File: embeddings/task/lightning_task/sequence_labeling.py [hunk 2]
```diff
@@ -33,11 +35,15 @@ def __init__(
...
         self.model = SequenceLabelingModule(
             model_name_or_path=self.model_name_or_path,
+            num_classes=self.num_classes,
             finetune_last_n_layers=self.finetune_last_n_layers,
             config_kwargs=self.model_config_kwargs,
             task_model_kwargs=self.task_model_kwargs,
         )
 
+    def restore_task_model(self, checkpoint_path: str) -> None:
+        self.model = SequenceLabelingModule.load_from_checkpoint(checkpoint_path)
```
Comment (ktagowski) [inline embeddings/task/lightning_task/sequence_labeling.py:45]: `classmethod` shouldn't be more unifed? And maybe just rename it to `load` ?

**Review by ktagowski — CHANGES_REQUESTED** [review]

**Comment by ktagowski** [comment]
<p>Added comments to https://app.reviewnb.com/CLARIN-PL/embeddings/blob/feature/trained-models-inference%2Fnotebooks%2Fvalidate_flair_models_inference.ipynb%2F/discussion</p>

**Comment by ktagowski** [comment]
<p>Added comments to <a href="https://app.reviewnb.com/CLARIN-PL/embeddings/blob/feature/trained-models-inference/notebooks/validate_lightning_models_inference.ipynb/" target="_blank">https://app.reviewnb.com/CLARIN-PL/embeddings/blob/feature/trained-models-inference/notebooks/validate_lightning_models_inference.ipynb/</a></p>

File: embeddings/model/lightning_module/huggingface_module.py (hunk 1 above)
Comment (djaniak) [reply embeddings/model/lightning_module/huggingface_module.py:28]: `load_from_checkpoint()` method does not work in another case - the model has to be initialized in the `__init__` for this method to work properly and read the saved checkpoint

File: embeddings/task/lightning_task/sequence_labeling.py (hunk 2 above)
Comment (djaniak) [reply embeddings/task/lightning_task/sequence_labeling.py:45]: on the second thought, I will remove these methods since we need to initialize the Task to use them so there is not much use for them anyway

Comment (djaniak) [reply embeddings/task/lightning_task/sequence_labeling.py:45]: we would need a clever way to initialize the task without the need for specifying the arguments in the `__init__`

File: embeddings/pipeline/lightning_sequence_labeling.py [hunk 3]
```diff
@@ -87,9 +87,11 @@ def __init__(
             load_dataset_kwargs=load_dataset_kwargs,
             **self.datamodule_kwargs
         )
+        datamodule.setup()
```
Comment (asawczyn) [inline embeddings/pipeline/lightning_sequence_labeling.py:90]: i do not like calling it manually. According to the warning it will be deprecated in the new version.

`pytorch_lightning/core/datamodule.py:469: LightningDeprecationWarning: DataModule.setup has already been called, so it will not be called again. In v1.6 this behavior will change to always call DataModule.setup.`

File: embeddings/task/lightning_task/text_classification.py [hunk 4]
```diff
@@ -21,18 +22,23 @@ def __init__(
...
     def build_task_model(self) -> None:
         self.model = TextClassificationModule(
             model_name_or_path=self.model_name_or_path,
+            num_classes=self.num_classes,
             finetune_last_n_layers=self.finetune_last_n_layers,
             config_kwargs=self.model_config_kwargs,
             task_model_kwargs=self.task_model_kwargs,
         )
 
+    def restore_task_model(self, checkpoint_path: str) -> None:
```
Comment (asawczyn) [inline embeddings/task/lightning_task/text_classification.py:39]: allow usage of Path instead of only str

**Review by asawczyn — CHANGES_REQUESTED** [review]

File: embeddings/pipeline/lightning_classification.py [hunk 5]
```diff
@@ -81,9 +81,11 @@ def __init__(
             load_dataset_kwargs=load_dataset_kwargs,
             **self.datamodule_kwargs
         )
+        datamodule.setup()
```
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/pipeline/lightning_classification.py:84]: We can either call it manually or ask user to give the num_classes argument to the pipelines since it is needed to initialize the lightning module

File: embeddings/pipeline/lightning_sequence_labeling.py (hunk 3 above)
Comment (djaniak) [reply embeddings/pipeline/lightning_sequence_labeling.py:89]: We can either call it manually or ask user to give the num_classes argument to the pipelines since it is needed to initialize the lightning module

File: embeddings/model/lightning_module/huggingface_module.py (hunk 1 above)
Comment (ktagowski) [reply embeddings/model/lightning_module/huggingface_module.py:28]: ok but I think's it needs better solution. I would keep this functions call in the `setup` and prepare appropiate configuration for stage  "prediction"? There is special hook that we can use while loading checkpoints  https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.core.hooks.CheckpointHooks.html#pytorch_lightning.core.hooks.CheckpointHooks.on_load_checkpoint

File: embeddings/task/lightning_task/sequence_labeling.py (hunk 2 above)
Comment (ktagowski) [reply embeddings/task/lightning_task/sequence_labeling.py]: I have added one idea in comment above that could handle this issue

File: embeddings/pipeline/lightning_sequence_labeling.py (hunk 3 above)
Comment (ktagowski) [reply embeddings/pipeline/lightning_sequence_labeling.py:89]: Same as above

File: notebooks/validate_flair_models_inference.ipynb [hunk 6]
```diff
@@ -0,0 +1,740 @@
+{
```
Comment (ktagowski) [inline notebooks/validate_flair_models_inference.ipynb:1]: It should be also moved to `tutorials`

File: notebooks/validate_flair_models_inference.ipynb [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+ "cells": [
+  {
+   "cell_type": "code",
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply notebooks/validate_flair_models_inference.ipynb:4]: Ohh, it is our main use case. It should be adapted to static embeddings

**Review by ktagowski — CHANGES_REQUESTED** [review]

File: embeddings/model/lightning_module/huggingface_module.py (hunk 1 above)
Comment (djaniak) [reply embeddings/model/lightning_module/huggingface_module.py:28]: I mean, the initialization of the model in PytorchLightning is usually done in __init__ method if you look at documentation so I don't think there is something wrong with this solution; moving initialization to setup was only done to not ask user to give the num classes arg

File: embeddings/task/lightning_task/sequence_labeling.py (hunk 2 above)
Comment (djaniak) [reply embeddings/task/lightning_task/sequence_labeling.py]: this was completely refactored in next commits so you can check how it works now

File: embeddings/model/lightning_module/huggingface_module.py (hunk 1 above)
Comment (ktagowski) [reply embeddings/model/lightning_module/huggingface_module.py:28]: I dissagree, I consider configuring models etc. as `setup` functionality than `init`

Comment (djaniak) [reply embeddings/model/lightning_module/huggingface_module.py:28]: Ok but if we using a framework such as `PytorchLightning` we should consider their good practices of designing/writing the code; as I said, if you look at any of the examples in the library the model is almost always initialized in the `__init__` - the setup is used only to build models dynamically or adjust something about them and thus it is not a default place of model building

Comment (djaniak) [reply embeddings/model/lightning_module/huggingface_module.py:28]: I really don't think we should reinvent their library inside of ours

File: embeddings/pipeline/lightning_classification.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/pipeline/lightning_classification.py:83]: https://github.com/PyTorchLightning/pytorch-lightning/issues/9943

we can do the workaround as described here

[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/pipeline/lightning_classification.py:83]: I would bump the Pytorch Lightning version at first, and then made addaptation / corrections

[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/pipeline/lightning_classification.py:83]: 1.6 also introduces new deprecations so It would be nice to check it now

File: embeddings/model/lightning_module/huggingface_module.py (hunk 1 above)
Comment (djaniak) [reply embeddings/model/lightning_module/huggingface_module.py:28]: Sorry I didn't mean to sound aggressive. 
1. It was placed in `__init__` at first but then it was moved to `setup()` to not force user to pass the `num_classes` argument. 
2. You are right but then we need to load the whole model with `on_load_checkpoint` method which I would call very unnecessary and undesirable. When moving it to `__init__` we allow for loading from checkpoint easily and stick to the convention.
3. Okay I can change the names to _init_model etc. if that seems as a better solution.

File: embeddings/pipeline/lightning_classification.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (asawczyn) [reply embeddings/pipeline/lightning_classification.py:83]: > So we should bump version in this PR and check whether it's still occurring.

It does not matter if it will occur. According to the message from warning it will be called many times which will be waste of time. 

I like the workaround. Instead of it, can we perform `setup()` on `init`?

[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/pipeline/lightning_classification.py:83]: Ok, after consideration of other issues, let's stick to 1.5.2 as for now. Sorry for that pushiness of 1.6, as I considered it as good moment for a bump as we haven't yet finished experiments. But considering other issues that would occur, the better solution is to stick with 1.5.2

[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/pipeline/lightning_classification.py:83]: > > So we should bump version in this PR and check whether it's still occurring.
> 
> It does not matter if it will occur. According to the message from warning it will be called many times which will be waste of time.
> 
> I like the workaround. Instead of it, can we perform `setup()` on `init`?

Absolutely, we load the dataset anyway in `__init__` to save the dataset info thus we can just perform it once and utilize loaded dataset info 

> Ok, after consideration of other issues, let's stick to 1.5.2 as for now. Sorry for that pushiness of 1.6, as I considered it as good moment for a bump as we haven't yet finished experiments. But considering other issues that would occur, the better solution is to stick with 1.5.2

Ok agreed

**Review by ktagowski — DISMISSED** [review]

File: tests/test_lightning_inference.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    lightning_classification_pipeline: LightningPipeline[
+        datasets.DatasetDict, Dict[str, np.ndarray], Dict[str, Any]
+    ],
+    tmp_path_module: "TemporaryDirectory[str]",
+) -> None:
+    pl.seed_everything(441, workers=True)
+    pipeline = lightning_classification_pipeline
+    results = pipeline.run()
+
+    ckpt_path = Path(tmp_path_module.name) / "checkpoints" / "epoch=0-step=1.ckpt"
```
Comment (asawczyn) [inline tests/test_lightning_inference.py:109]:
```suggestion
    ckpt_path = tmp_path_module / "checkpoints" / "epoch=0-step=1.ckpt"
```

File: tests/test_lightning_inference.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        accelerator="cpu",
+        **dataset_kwargs,
+    )
+
+
+def test_lightning_pipeline_inference(
+    lightning_classification_pipeline: LightningPipeline[
+        datasets.DatasetDict, Dict[str, np.ndarray], Dict[str, Any]
+    ],
+    tmp_path_module: "TemporaryDirectory[str]",
```
Comment (asawczyn) [inline tests/test_lightning_inference.py:103]:
```suggestion
    tmp_path_module: Path,
```

File: tests/test_lightning_inference.py (hunk 8 above)
Comment (asawczyn) [inline tests/test_lightning_inference.py:109]: Is it the only one saved checkpoint? I would prefer "best.ckpt" or similar

File: tests/test_lightning_inference.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+) -> None:
+    pl.seed_everything(441, workers=True)
+    pipeline = lightning_classification_pipeline
+    results = pipeline.run()
+
+    ckpt_path = Path(tmp_path_module.name) / "checkpoints" / "epoch=0-step=1.ckpt"
+    task_from_ckpt = TextClassificationTask.from_checkpoint(
+        checkpoint_path=ckpt_path.resolve(),
+        output_path=tmp_path_module.name,
```
Comment (asawczyn) [inline tests/test_lightning_inference.py:112]:
```suggestion
        output_path=tmp_path_module,
```

File: tests/test_lightning_inference.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    )
+
+    model_state_dict = pipeline.model.task.model.model.state_dict()
+    model_from_ckpt_state_dict = task_from_ckpt.model.model.state_dict()
+    assert model_state_dict.keys() == model_from_ckpt_state_dict.keys()
+    for k in model_state_dict.keys():
+        assert torch.equal(model_state_dict[k], model_from_ckpt_state_dict[k])
+
+    test_dataloader = pipeline.datamodule.test_dataloader()
+    predictions = task_from_ckpt.predict(test_dataloader, return_names=False)
```
Comment (asawczyn) [inline tests/test_lightning_inference.py:122]: Is it true that  `return_names=False` is caused by unavailability of train dataset? If it is true, names have to be saved to hparams. Receiving array of probabilities does not make sense, if we do not know how to map index to label.

File: tests/test_lightning_inference.py (hunk 9 above)
Comment (asawczyn) [inline tests/test_lightning_inference.py:103]: can we append these tests to `tests/test_lightning_classification_pipeline.py` where the model is already trained? It will speed up tests.

File: tests/test_flair_inference.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    transformed_data = pipeline.transformation.transform(loaded_data)
+    test_data = transformed_data.test
+
+    y_pred, loss = task_from_ckpt.predict(test_data)
+    y_true = task_from_ckpt.get_y(test_data, task_from_ckpt.y_type, task_from_ckpt.y_dictionary)
+    results_from_ckpt = pipeline.evaluator.evaluate({"y_pred": y_pred, "y_true": y_true})
+    assert np.array_equal(result["data"]["y_pred"], results_from_ckpt["data"]["y_pred"])
+
+
+def test_sequence_labeling_inference(
```
Comment (asawczyn) [inline tests/test_flair_inference.py:122]: can we append these tests to `tests/test_flair..` where the model is already trained? It will speed up tests.

File: tutorials/validate_lightning_models_inference.ipynb [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ee9e824c-00f1-45b0-9e32-1bd33f364f3a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ckpt_path = output_path / \"checkpoints\" / \"epoch=0-step=1.ckpt\"\n",
```
Comment (asawczyn) [inline tutorials/validate_lightning_models_inference.ipynb:168]: as above, should be `best`

**Review by asawczyn — CHANGES_REQUESTED** [review]

File: tests/test_lightning_inference.py (hunk 8 above)
Comment (djaniak) [reply tests/test_lightning_inference.py:109]: If we set `best.ckpt` name then it will be ambiguous in case we would want to save top_k checkpoints. You can format the name in a way specified here: https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.callbacks.ModelCheckpoint.html
and you could also add a flag to save the last checkpoint as `last.ckpt`

File: tests/test_lightning_inference.py (hunk 11 above)
Comment (djaniak) [reply tests/test_lightning_inference.py:122]: It is caused by unavailability of datamodule (we pass only dataloader for prediction). I'll save the names from datamodule to hparams method.

File: tests/test_lightning_inference.py (hunk 9 above)
Comment (djaniak) [reply tests/test_lightning_inference.py]: yes

File: tests/test_flair_inference.py (hunk 12 above)
Comment (djaniak) [reply tests/test_flair_inference.py]: ok

File: embeddings/model/lightning_module/sequence_labeling.py [hunk 14]
```diff
@@ -73,3 +88,64 @@ def test_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
         else:
             _logger.warning("Missing labels for the test data")
         return None
+
+    def str2int(self, values: Union[str, Iterable[Any]]) -> Union[int, Iterable[Any]]:
+        """Conversion class name string => integer duplicated from huggingface ClassLabel."""
```
Comment (ktagowski) [inline embeddings/model/lightning_module/sequence_labeling.py:93]: I'm not a big fan of copying functionality from other libs. Why we can't just use wrap ClassLabel object usage?

**Review by ktagowski — CHANGES_REQUESTED** [review]

**Review by asawczyn — CHANGES_REQUESTED** [review]

**Review by asawczyn — APPROVED** [review]
